File: packages/gpu-view/src/features.ts
/**
 * Sparse feature rows per model token. Must match training/gpu_view/features.py byte
 * for byte (fixtures in test/fixtures/features.json).
 *
 * The model never sees a field name. Schema membership arrives as anonymous rows:
 * "matched a field of kind K (begin/inside, exact/stem/prefix/typo, via an alias)",
 * "matched an enum value (unique owner? owned by the nearest preceding/following
 * field?)", and the kind of / distance to the nearest field match on either side.
 * The only exact word identities are the closed task lexicon (lexicon.ts); everything
 * else is a hashed bucket (word + consonant skeleton), shape and length. Whitespace
 * tokens are dropped before the model.
 */
import {
	CharClass,
	type FeatureRows,
	hashToken,
	Shape,
} from "@gpu-utils/runtime";
import { KEYWORD_COUNT, KEYWORD_ID } from "./lexicon.ts";
import {
	enumEntries,
	type FieldKind,
	fieldEntries,
	matchSpans,
	modelTokens,
	type Schema,
	type Span,
} from "./match.ts";

const KIND_ID: Record<FieldKind, number> = {
	text: 1,
	number: 2,
	date: 3,
	enum: 4,
	boolean: 5,
};
const LENGTH_BUCKETS = [1, 2, 3, 4, 6, 8, 12];
const WORD_BUCKETS = 128;
const SKEL_BUCKETS = 64;
const DIST_ROWS = 6;

export const BLOCKS: [string, number][] = [
	["shape", 8],
	["length", 8],
	["word", WORD_BUCKETS],
	["skeleton", SKEL_BUCKETS],
	["keyword", KEYWORD_COUNT + 1],
	["flags", 10],
	["field_kind", 6],
	["field_pos", 3],
	["quality", 5],
	["alias", 2],
	["enum_any", 2],
	["enum_pos", 3],
	["enum_unique", 2],
	["enum_prev", 2],
	["enum_next", 2],
	["prev_kind", 6],
	["prev_dist", DIST_ROWS],
	["next_kind", 6],
	["next_dist", DIST_ROWS],
	["position", 4],
];
export const OFFSET: Record<string, number> = {};
let cursor = 0;
for (const [name, size] of BLOCKS) {
	OFFSET[name] = cursor;
	cursor += size;
}
export const FEATURE_ROWS = cursor;
export const PADDING_ROW = FEATURE_ROWS;
export const SLOTS = 28;

const SUFFIXES = new Set(["k", "m", "b", "bn", "mm", "%"]);
const DIGIT = /\p{Nd}/u;

function lengthBucket(n: number): number {
	for (let i = 0; i < LENGTH_BUCKETS.length; i++)
		if (n <= LENGTH_BUCKETS[i]!) return i;
	return LENGTH_BUCKETS.length;
}

function skeleton(text: string): string {
	const s = text.toLowerCase().replace(/[aeiou]/g, "");
	return s || text.toLowerCase();
}

const distBucket = (d: number | null) =>
	d === null ? DIST_ROWS - 1 : Math.min(d, DIST_ROWS - 2);

export interface ViewFeatures extends FeatureRows {
	fieldSpans: Span[];
	enumSpans: Span[];
}

export function featurize(text: string, schema: Schema): ViewFeatures {
	const tokens = modelTokens(text);
	const n = tokens.length;
	const fields = matchSpans(tokens, fieldEntries(schema));
	const enums = matchSpans(tokens, enumEntries(schema));
	const rows: number[][] = [];
	if (n === 0) return { tokens, rows, fieldSpans: fields, enumSpans: enums };

	const fieldAt: (Span | null)[] = new Array(n).fill(null);
	for (const s of fields) for (let i = s.start; i < s.end; i++) fieldAt[i] = s;
	const enumAt: (Span | null)[] = new Array(n).fill(null);
	for (const s of enums) for (let i = s.start; i < s.end; i++) enumAt[i] = s;

	const prevSpan: (Span | null)[] = new Array(n).fill(null);
	const nextSpan: (Span | null)[] = new Array(n).fill(null);
	let last: Span | null = null;
	for (let i = 0; i < n; i++) {
		const own = fieldAt[i]!;
		prevSpan[i] = own === null || last !== own ? last : before(fields, own);
		if (own !== null) last = own;
	}
	let upcoming: Span | null = null;
	for (let i = n - 1; i >= 0; i--) {
		const own = fieldAt[i]!;
		nextSpan[i] =
			own === null || upcoming !== own ? upcoming : after(fields, own);
		if (own !== null) upcoming = own;
	}

	for (let i = 0; i < n; i++) {
		const tok = tokens[i]!;
		const low = tok.text.toLowerCase();
		const r: number[] = [];
		const add = (block: string, v: number) => r.push(OFFSET[block]! + v);

		add("shape", tok.shape);
		add("length", lengthBucket(Array.from(tok.text).length)); // code points, like Python len()
		add("word", hashToken(tok.text, WORD_BUCKETS));
		add("skeleton", hashToken(skeleton(tok.text), SKEL_BUCKETS));
		add("keyword", KEYWORD_ID.get(low) ?? KEYWORD_COUNT);

		if (DIGIT.test(tok.text)) add("flags", 0);
		if (tok.cls === CharClass.Digit) add("flags", 1);
		if (tok.cls === CharClass.Other) add("flags", 2);
		if (tok.shape === Shape.Title || tok.shape === Shape.Upper) add("flags", 3);
		if (i === 0) add("flags", 4);
		if (i === n - 1) add("flags", 5);
		const prevDigit = i > 0 && tokens[i - 1]!.cls === CharClass.Digit;
		if (prevDigit) add("flags", 6);
		if (i + 1 < n && tokens[i + 1]!.cls === CharClass.Digit) add("flags", 7);
		if (
			tok.cls === CharClass.Digit &&
			tok.text.length === 4 &&
			(tok.text.startsWith("19") || tok.text.startsWith("20"))
		)
			add("flags", 8);
		if (prevDigit && SUFFIXES.has(low)) add("flags", 9);

		const f = fieldAt[i]!;
		add("field_kind", f ? KIND_ID[f.kind] : 0);
		add("field_pos", f === null ? 0 : i === f.start ? 1 : 2);
		add("quality", f ? f.quality + 1 : 0);
		add("alias", f?.alias ? 1 : 0);

		const e = enumAt[i]!;
		const p = prevSpan[i]!;
		const q = nextSpan[i]!;
		add("enum_any", e ? 1 : 0);
		add("enum_pos", e === null ? 0 : i === e.start ? 1 : 2);
		add("enum_unique", e && e.owners.length === 1 ? 1 : 0);
		add("enum_prev", e && p !== null && e.owners.includes(p.field) ? 1 : 0);
		add("enum_next", e && q !== null && e.owners.includes(q.field) ? 1 : 0);

		add("prev_kind", p ? KIND_ID[p.kind] : 0);
		add("prev_dist", distBucket(p ? i - p.end : null));
		add("next_kind", q ? KIND_ID[q.kind] : 0);
		add("next_dist", distBucket(q ? q.start - i - 1 : null));
		add("position", Math.min(3, Math.floor((4 * i) / n)));
		rows.push(r);
	}
	return { tokens, rows, fieldSpans: fields, enumSpans: enums };
}

function before(spans: Span[], own: Span): Span | null {
	let best: Span | null = null;
	for (const s of spans) if (s.end <= own.start) best = s;
	return best;
}

function after(spans: Span[], own: Span): Span | null {
	for (const s of spans) if (s.start >= own.end) return s;
	return null;
}

File: packages/gpu-view/training/gpu_view/features.py
"""Sparse feature rows per model token. Must match src/features.ts byte for byte.

The model never sees a field name. Schema membership arrives as anonymous rows:
"this token matched a field of kind K (begin/inside, exact/stem/prefix/typo, via
an alias)", "this token matched an enum value (unique owner? owned by the nearest
preceding/following field?)", and the kind of / distance to the nearest field
match on either side. The only exact word identities are the closed task lexicon;
everything else is a hashed bucket (word + consonant skeleton), shape and length.

Whitespace tokens are dropped before the model: the sequence the model sees is
`match.model_tokens(text)`.
"""

from __future__ import annotations

import unicodedata

from gpu_utils_training.features import CLASS_DIGIT, CLASS_OTHER, SHAPE_TITLE, SHAPE_UPPER, Token, hash_token

from . import match
from .lexicon import KEYWORD_COUNT, KEYWORD_ID

KIND_ID = {"text": 1, "number": 2, "date": 3, "enum": 4, "boolean": 5}
LENGTH_BUCKETS = [1, 2, 3, 4, 6, 8, 12]
WORD_BUCKETS = 128
SKEL_BUCKETS = 64
DIST_ROWS = 6  # 0,1,2,3,4+,none

BLOCKS: list[tuple[str, int]] = [
    ("shape", 8),
    ("length", 8),
    ("word", WORD_BUCKETS),
    ("skeleton", SKEL_BUCKETS),
    ("keyword", KEYWORD_COUNT + 1),
    ("flags", 12),
    ("field_kind", 6),
    ("field_pos", 3),
    ("quality", 6),
    ("alias", 2),
    ("field_neg", 2),
    ("enum_any", 2),
    ("enum_pos", 3),
    ("enum_unique", 2),
    ("enum_prev", 2),
    ("enum_next", 2),
    ("prev_kind", 6),
    ("prev_dist", DIST_ROWS),
    ("next_kind", 6),
    ("next_dist", DIST_ROWS),
    ("position", 4),
]
OFFSET: dict[str, int] = {}
_cursor = 0
for _name, _size in BLOCKS:
    OFFSET[_name] = _cursor
    _cursor += _size
FEATURE_ROWS = _cursor
PADDING_ROW = FEATURE_ROWS
SLOTS = 29

FLAG_HAS_DIGIT, FLAG_ALL_DIGIT, FLAG_PUNCT, FLAG_UPPER, FLAG_FIRST, FLAG_LAST = 0, 1, 2, 3, 4, 5
FLAG_PREV_DIGIT, FLAG_NEXT_DIGIT, FLAG_YEAR, FLAG_SUFFIX, FLAG_ER, FLAG_EST = 6, 7, 8, 9, 10, 11


def length_bucket(n: int) -> int:
    for i, b in enumerate(LENGTH_BUCKETS):
        if n <= b:
            return i
    return len(LENGTH_BUCKETS)


def skeleton(text: str) -> str:
    s = "".join(c for c in text.lower() if c not in "aeiou")
    return s or text.lower()


def dist_bucket(d: int | None) -> int:
    if d is None:
        return DIST_ROWS - 1
    return min(d, DIST_ROWS - 2)


def featurize(text: str, schema: dict) -> tuple[list[Token], list[list[int]]]:
    tokens = match.model_tokens(text)
    n = len(tokens)
    if n == 0:
        return tokens, []
    fields = match.match_spans(tokens, match.field_entries(schema))
    enums = match.match_spans(tokens, match.enum_entries(schema), enum=True)

    field_at: list[match.Span | None] = [None] * n
    for s in fields:
        for i in range(s.start, s.end):
            field_at[i] = s
    enum_at: list[match.Span | None] = [None] * n
    for s in enums:
        for i in range(s.start, s.end):
            enum_at[i] = s

    # Nearest field span strictly before / after each token (excluding its own span).
    prev_span: list[match.Span | None] = [None] * n
    next_span: list[match.Span | None] = [None] * n
    last: match.Span | None = None
    for i in range(n):
        own = field_at[i]
        prev_span[i] = last if (own is None or last is not own) else _before(fields, own)
        if own is not None:
            last = own
    upcoming: match.Span | None = None
    for i in range(n - 1, -1, -1):
        own = field_at[i]
        next_span[i] = upcoming if (own is None or upcoming is not own) else _after(fields, own)
        if own is not None:
            upcoming = own

    rows: list[list[int]] = []
    for i, tok in enumerate(tokens):
        low = tok.text.lower()
        r: list[int] = []

        def add(block: str, v: int) -> None:
            r.append(OFFSET[block] + v)

        add("shape", tok.shape)
        add("length", length_bucket(len(tok.text)))
        add("word", hash_token(tok.text, WORD_BUCKETS))
        add("skeleton", hash_token(skeleton(tok.text), SKEL_BUCKETS))
        add("keyword", KEYWORD_ID.get(low, KEYWORD_COUNT))

        has_digit = any(unicodedata.category(c) == "Nd" for c in tok.text)
        if has_digit:
            add("flags", FLAG_HAS_DIGIT)
        if tok.cls == CLASS_DIGIT:
            add("flags", FLAG_ALL_DIGIT)
        if tok.cls == CLASS_OTHER:
            add("flags", FLAG_PUNCT)
        if tok.shape in (SHAPE_TITLE, SHAPE_UPPER):
            add("flags", FLAG_UPPER)
        if i == 0:
            add("flags", FLAG_FIRST)
        if i == n - 1:
            add("flags", FLAG_LAST)
        if i > 0 and tokens[i - 1].cls == CLASS_DIGIT:
            add("flags", FLAG_PREV_DIGIT)
        if i + 1 < n and tokens[i + 1].cls == CLASS_DIGIT:
            add("flags", FLAG_NEXT_DIGIT)
        if tok.cls == CLASS_DIGIT and len(tok.text) == 4 and tok.text[:2] in ("19", "20"):
            add("flags", FLAG_YEAR)
        if i > 0 and tokens[i - 1].cls == CLASS_DIGIT and low in ("k", "m", "b", "bn", "mm", "%"):
            add("flags", FLAG_SUFFIX)

        if tok.cls == 0 and len(low) >= 5 and (low.endswith("er") or low.endswith("ier")):
            add("flags", FLAG_ER)
        if tok.cls == 0 and len(low) >= 6 and (low.endswith("est") or low.endswith("iest")):
            add("flags", FLAG_EST)

        f = field_at[i]
        add("field_kind", KIND_ID[f.kind] if f else 0)
        add("field_pos", 0 if f is None else (1 if i == f.start else 2))
        add("quality", f.quality + 1 if f else 0)
        add("alias", 1 if (f and f.alias) else 0)
        add("field_neg", 1 if (f and f.neg) else 0)

        e = enum_at[i]
        add("enum_any", 1 if e else 0)
        add("enum_pos", 0 if e is None else (1 if i == e.start else 2))
        add("enum_unique", 1 if (e and len(e.owners) == 1) else 0)
        p, q = prev_span[i], next_span[i]
        add("enum_prev", 1 if (e and p is not None and p.field in e.owners) else 0)
        add("enum_next", 1 if (e and q is not None and q.field in e.owners) else 0)

        add("prev_kind", KIND_ID[p.kind] if p else 0)
        add("prev_dist", dist_bucket(i - p.end if p else None))
        add("next_kind", KIND_ID[q.kind] if q else 0)
        add("next_dist", dist_bucket(q.start - i - 1 if q else None))
        add("position", min(3, (4 * i) // n))
        assert len(r) <= SLOTS
        rows.append(r)
    return tokens, rows


def _before(spans: list[match.Span], own: match.Span) -> match.Span | None:
    best = None
    for s in spans:
        if s.end <= own.start:
            best = s
    return best


def _after(spans: list[match.Span], own: match.Span) -> match.Span | None:
    for s in spans:
        if s.start >= own.end:
            return s
    return None


def pad(rows: list[list[int]]) -> list[list[int]]:
    return [r[:SLOTS] + [PADDING_ROW] * (SLOTS - len(r)) for r in rows]


if __name__ == "__main__":
    print(f"FEATURE_ROWS = {FEATURE_ROWS}, SLOTS = {SLOTS}")
    for name, size in BLOCKS:
        print(f"  {name:<12} {OFFSET[name]:>4} .. {OFFSET[name] + size - 1:>4}")
    schema = {"fields": [
        {"name": "customer", "kind": "text"},
        {"name": "country", "kind": "enum", "values": ["Germany", "France"]},
        {"name": "orders", "kind": "number", "aliases": ["order count"]},
        {"name": "created_at", "kind": "date", "aliases": ["created", "signed up"]},
    ]}
    toks, rows = featurize("customers in Germany or France with more than 5 orders", schema)
    for t, r in zip(toks, rows):
        named = []
        for v in r:
            for name, size in BLOCKS:
                if OFFSET[name] <= v < OFFSET[name] + size:
                    named.append(f"{name}+{v - OFFSET[name]}")
                    break
        print(f"  {t.text:<10} {' '.join(named)}")
